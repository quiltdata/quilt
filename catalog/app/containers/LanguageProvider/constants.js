/*
 *
 * LanguageProvider constants
 *
 */

export const REDUX_KEY = 'app/LanguageProvider';

export const CHANGE_LOCALE = 'app/LanguageProvider/CHANGE_LOCALE';
export const DEFAULT_LOCALE = 'en';
