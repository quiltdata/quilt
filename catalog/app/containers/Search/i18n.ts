// Descriptive filter labels
export const PACKAGE_FILTER_LABELS = {
  bucket: 'Bucket',
  comment: 'Package revision comment',
  entries: 'Total number of files in the package',
  hash: 'Package hash',
  modified: 'Last modified',
  name: 'Package name',
  size: 'Cumulative package size',
  workflow: 'Workflow',
}

// Shortened filter labels
export const COLUMN_LABELS = {
  bucket: 'Bucket',
  comment: 'Comment',
  entries: 'File count',
  hash: 'Hash',
  modified: 'Modified',
  name: 'Package',
  size: 'Size',
  workflow: 'Workflow',
}

export const OBJECT_FILTER_LABELS = {
  content: 'Contents',
  deleted: 'Delete marker',
  ext: 'Extension',
  key: 'Object key',
  modified: 'Last modified',
  size: 'Object size',
}
