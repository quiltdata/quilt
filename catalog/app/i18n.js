/**
 * i18n.js
 *
 * This will setup the i18n language files and locale data for your app.
 *
 *   IMPORTANT: This file is used by the internal build
 *   script `extract-intl`, and must use CommonJS module syntax
 *   You CANNOT use import/export in this file.
 */
const enTranslationMessages = require('./translations/en.json')

const DEFAULT_LOCALE = 'en'
exports.DEFAULT_LOCALE = DEFAULT_LOCALE

const appLocales = ['en']
exports.appLocales = appLocales

const formatTranslationMessages = (locale, messages) => {
  const defaultFormattedMessages =
    locale !== DEFAULT_LOCALE
      ? formatTranslationMessages(DEFAULT_LOCALE, enTranslationMessages)
      : {}
  return Object.keys(messages).reduce((formattedMessages, key) => {
    let message = messages[key]
    if (!message && locale !== DEFAULT_LOCALE) {
      message = defaultFormattedMessages[key]
    }
    return Object.assign(formattedMessages, { [key]: message })
  }, {})
}
exports.formatTranslationMessages = formatTranslationMessages

const translationMessages = {
  en: formatTranslationMessages('en', enTranslationMessages),
}
exports.translationMessages = translationMessages
