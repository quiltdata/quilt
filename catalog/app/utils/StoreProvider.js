import PT from 'prop-types';
import React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import { setPropTypes } from 'recompose';

import { composeComponent } from 'utils/reactTools';
import { ReducerInjector } from 'utils/ReducerInjector';
import { SagaInjector } from 'utils/SagaInjector';

export default composeComponent('StoreProvider',
  setPropTypes({
    store: PT.object.isRequired,
  }),
  ({ store, children }) => (
    <ReduxProvider store={store}>
      <ReducerInjector inject={store.injectReducer}>
        <SagaInjector run={store.runSaga}>
          {children}
        </SagaInjector>
      </ReducerInjector>
    </ReduxProvider>
  ));
