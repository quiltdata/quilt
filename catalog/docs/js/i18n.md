# `i18n`

`react-intl` is a library to manage internationalization and pluralization support
for your react application. This involves multi-language support for both the static text but also things like variable numbers, words or names that change with application state. `react-intl` provides an incredible amount of mature facility to preform these very tasks.

The complete `react-intl` docs can be found here:

https://github.com/yahoo/react-intl/wiki

## Usage

Below we see a `messages.js` file for the `Footer` component example. A `messages.js` file should be included in any simple or container component that wants to use internationalization. You can add this support when you scaffold your component using this boilerplates scaffolding `plop` system.

All default English text for the component is contained here (e.g. `This project is licensed under the MIT license.`), and is tagged with an ID (e.g. `boilerplate.components.Footer.license.message`) in addition to it's object definition id (e.g. `licenseMessage`).

This is set in `react-intl`'s `defineMessages` function which is then exported for use in the component. You can read more about `defineMessages` here:

https://github.com/yahoo/react-intl/wiki/API#definemessages

```js
/*
 * Footer Messages
 *
 * This contains all the text for the Footer component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  licenseMessage: {
    id: 'boilerplate.components.Footer.license.message',
    defaultMessage: 'This project is licensed under the MIT license.',
  },
  authorMessage: {
    id: 'boilerplate.components.Footer.author.message',
    defaultMessage: `
      Made with love by {author}.
    `,
  },
});
```

Below is the example `Footer` component. Here we see the component including the `messages.js` file, which contains all the default component text, organized with ids (and optionally descriptions). We are also importing the `FormattedMessage` component, which will display a given message from the `messages.js` file in the selected language.

You will also notice a more complex use of `FormattedMessage` for the author message where alternate or variable values (i.e. `author: <A href="https://twitter.com/mxstbr">Max Stoiber</A>,`) are being injected, in this case it's a react component.

```js
import React from 'react';

import messages from './messages';
import A from 'components/A';
import styles from './styles.css';
import { FormattedMessage } from 'react-intl';

function Footer() {
  return (
    <footer className={styles.footer}>
      <section>
        <p>
          <FormattedMessage {...messages.licenseMessage} />
        </p>
      </section>
      <section>
        <p>
          <FormattedMessage
            {...messages.authorMessage}
            values={{
              author: <A href="https://twitter.com/mxstbr">Max Stoiber</A>,
            }}
          />
        </p>
      </section>
    </footer>
  );
}

export default Footer;
```

## Extracting i18n JSON files

You can extract all i18n language within each component by running the following command:

```
  npm run extract-intl
```

This will extract all language into i18n JSON files in `app/translations`.

## Adding A Language

You can add a language by running the generate command:

```
  npm run generate language
```

Then enter the two character i18n standard language specifier (e.g. "fr", "de", "es" - without quotes). This will add in the necessary JSON language file and import statements for the language. Note, it is up to you to fill in the translations for the language.

## Removing i18n and react-intl

You can remove `react-intl` modules by first removing the `IntlProvider` object from the `app/app.js` file and by either removing or not selecting the i18n text option during component scaffolding.

The packages associated with `react-intl` are:
 - react-intl
 - babel-plugin-react-intl
