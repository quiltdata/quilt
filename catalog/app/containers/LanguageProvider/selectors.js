import { createSelector } from 'reselect';

import { REDUX_KEY } from './constants';

/**
 * Direct selector to the languageToggle state domain
 */
const selectLanguage = (state) => state.get(REDUX_KEY);

/**
 * Select the language locale
 */

const makeSelectLocale = () => createSelector(
  selectLanguage,
  (languageState) => languageState.get('locale')
);

export {
  selectLanguage,
  makeSelectLocale,
};
