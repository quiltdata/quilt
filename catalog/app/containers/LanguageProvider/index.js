/*
 *
 * LanguageProvider
 *
 * this component connects the redux state language locale to the
 * IntlProvider component and i18n messages (loaded from `app/translations`)
 */

import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { mapProps, setPropTypes } from 'recompose';
import { createSelector } from 'reselect';
import { IntlProvider } from 'react-intl';

import { composeComponent } from 'utils/reactTools';
import { injectReducer } from 'utils/ReducerInjector';

import { REDUX_KEY } from './constants';
import reducer from './reducer';
import { makeSelectLocale } from './selectors';

export default composeComponent('LanguageProvider',
  injectReducer(REDUX_KEY, reducer),
  connect(createSelector(
    makeSelectLocale(),
    (locale) => ({ locale })
  )),
  setPropTypes({
    locale: PropTypes.string,
    messages: PropTypes.object,
    children: PropTypes.element.isRequired,
  }),
  mapProps(({ locale, messages, ...props }) => ({
    locale,
    key: locale,
    messages: messages[locale],
    ...props,
  })),
  IntlProvider);
