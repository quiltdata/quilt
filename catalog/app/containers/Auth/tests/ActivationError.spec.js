import createHistory from 'history/createMemoryHistory';
import { fromJS } from 'immutable';
import React from 'react';

import LanguageProvider from 'containers/LanguageProvider';
import { translationMessages as messages } from 'i18n';
import StoreProvider from 'utils/StoreProvider';
import configureStore from 'store';

import feature from 'testing/feature';
import reactSteps from 'testing/react';

import { ActivationError } from '..';

jest.mock('constants/config');


const setup = () => {
  const history = createHistory({ initialEntries: ['/'] });
  const store = configureStore(fromJS({}), history);
  const tree = (
    <StoreProvider store={store}>
      <LanguageProvider messages={messages}>
        <ActivationError />
      </LanguageProvider>
    </StoreProvider>
  );
  return { tree };
};

const screens = {
  error: (html) => {
    expect(html.find('h1').text()).toBe('Activation Error');
    expect(html.find('p').text()).toMatch(/Something.+activation[^]+support@quiltdata\.io/m);
  },
};

const steps = [
  ...reactSteps({ setup, screens }),
];


feature('containers/Auth/ActivationError')
  .scenario('Mounting the component')

  .when('the component tree is mounted')
  .then('I should see the error screen')
  .then('the rendered markup should match the snapshot')

  .steps(steps)
  .run();
