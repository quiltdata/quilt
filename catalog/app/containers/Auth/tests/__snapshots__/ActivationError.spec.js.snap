// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature: containers/Auth/ActivationError Background: Scenario: Mounting the component When the component tree is mounted Then the rendered markup should match the snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  max-width: 280px;
  min-height: calc(100vh - 300px);
  width: 100%;
}

.c1 {
  text-align: center;
}

.c2 {
  text-align: center;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
    <span>
      Activation Error
    </span>
  </h1>
  <p
    class="c2"
  >
    <span>
      Something went wrong during account activation.
      We're here to help. Email us at 
      <a
        href="mailto:support@quiltdata.io"
      >
        support@quiltdata.io
      </a>
      .
    </span>
  </p>
</div>
`;
